pkg/services/ngalert/schedule/schedule.go

```diff
@@ -424,84 +422,110 @@ func (sch *schedule) ruleEvaluationLoop(ctx context.Context) error {
+		dur := sch.clock.Now().Sub(start)
+		evalTotalMetric.Inc()
+		evalDurationMetric.Observe(dur.Seconds())
+		if err != nil {
+			evalTotalFailuresMetric.Inc()
+			// consider saving alert instance on error
+			logger.Error("failed to evaluate alert rule", "duration", dur, "error", err)
+			return err
+		}
+		logger.Debug("alert rule evaluated", "results", results, "duration", dur)
 
-				processedStates := sch.stateManager.ProcessEvalResults(context.Background(), alertRule, results)
-				sch.saveAlertStates(processedStates)
-				alerts := FromAlertStateToPostableAlerts(processedStates, sch.stateManager, sch.appURL)
+		processedStates := sch.stateManager.ProcessEvalResults(context.Background(), alertRule, results)
+		sch.saveAlertStates(processedStates)
+		alerts := FromAlertStateToPostableAlerts(processedStates, sch.stateManager, sch.appURL)
 
-				if len(alerts.PostableAlerts) == 0 {
-					sch.log.Debug("no alerts to put in the notifier", "org", alertRule.OrgID)
-					return nil
-				}
+		if len(alerts.PostableAlerts) == 0 {
+			logger.Debug("no alerts to put in the notifier")
+			return nil
+		}
 
-				sch.log.Debug("sending alerts to notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts, "org", alertRule.OrgID)
-				n, err := sch.multiOrgNotifier.AlertmanagerFor(alertRule.OrgID)
-				if err == nil {
-					if err := n.PutAlerts(alerts); err != nil {
-						sch.log.Error("failed to put alerts in the notifier", "count", len(alerts.PostableAlerts), "err", err)
-					}
-				} else {
-					sch.log.Error("unable to lookup local notifier for this org - alerts not delivered", "org", alertRule.OrgID, "count", len(alerts.PostableAlerts), "err", err)
-				}
+		var localNotifierExist, externalNotifierExist bool
+		logger.Debug("sending alerts to notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
+		n, err := sch.multiOrgNotifier.AlertmanagerFor(alertRule.OrgID)
+		if err == nil {
+			localNotifierExist = true
+			if err := n.PutAlerts(alerts); err != nil {
+				logger.Error("failed to put alerts in the local notifier", "count", len(alerts.PostableAlerts), "error", err)
+			}
+		} else {
+			if errors.Is(err, notifier.ErrNoAlertmanagerForOrg) {
+				logger.Debug("local notifier was not found")
+			} else {
+				logger.Error("local notifier is not available", "error", err)
+			}
+		}
 
-				// Send alerts to external Alertmanager(s) if we have a sender for this organization.
-				sch.sendersMtx.RLock()
-				defer sch.sendersMtx.RUnlock()
-				s, ok := sch.senders[alertRule.OrgID]
-				if ok {
-					s.SendAlerts(alerts)
-				}
+		// Send alerts to external Alertmanager(s) if we have a sender for this organization.
+		sch.sendersMtx.RLock()
+		defer sch.sendersMtx.RUnlock()
+		s, ok := sch.senders[alertRule.OrgID]
+		if ok {
+			logger.Debug("found external notifier. Sending alerts")
+			s.SendAlerts(alerts)
+			externalNotifierExist = true
+		} else {
+			logger.Debug("external notifier was not found")
```

Comment:
This line seems unnecessary, we know that if the above line was not present we didn't have any notifiers to send to. Perhaps we should even find a way to not take the lock entirely - but probably part of a separate PR.